In [1]:
import pandas

mydataset = {
  'cars': ["BMW", "Volvo", "Ford"],
  'passings': [3, 7, 2]
}

myvar = pandas.DataFrame(mydataset)

print(myvar)

    cars  passings
0    BMW         3
1  Volvo         7
2   Ford         2


In [2]:
import pandas as pd

mydataset = {
  'cars': ["BMW", "Volvo", "Ford"],
  'passings': [3, 7, 2]
}

myvar = pd.DataFrame(mydataset)

print(myvar)

    cars  passings
0    BMW         3
1  Volvo         7
2   Ford         2


In [3]:
import pandas as pd

print(pd.__version__)

2.2.3


# Pandas Series
1차원 배열

In [7]:
import pandas as pd

a = [1, 7, 2]

myvar = pd.Series(a)

print(myvar)

0    1
1    7
2    2
dtype: int64


## Labels

In [8]:
print(myvar[0])
print(myvar[1])
print(myvar[2])

1
7
2


## Create Labels

In [9]:
import pandas as pd

a = [1, 7, 2]

myvar = pd.Series(a, index = ["x", "y", "z"])

print(myvar)

x    1
y    7
z    2
dtype: int64


In [10]:
print(myvar["x"])
print(myvar["y"])
print(myvar["z"])

1
7
2


## Key/Value Objects as Series
딕셔너리의 키가 라벨이 된다.

In [11]:
import pandas as pd

calories = {"day1": 420, "day2": 380, "day3": 390}

myvar = pd.Series(calories)

print(myvar)

day1    420
day2    380
day3    390
dtype: int64


In [14]:
import pandas as pd

calories = {"day1": 420, "day2": 380, "day3": 390}

myvar = pd.Series(calories, index = ["day1", "day2"])

print(myvar)

day1    420
day2    380
dtype: int64


# DataFrame
다차원 테이블
- series는 열과 같고, DataFrame은 전체 테이블

In [16]:
import pandas as pd

data = {
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}

myvar = pd.DataFrame(data)

print(myvar)
myvar

   calories  duration
0       420        50
1       380        40
2       390        45


calories  duration
0       420        50
1       380        40
2       390        45

In [18]:
import pandas as pd

data = {
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}

#load data into a DataFrame object:
df = pd.DataFrame(data)

print(df) 
df

   calories  duration
0       420        50
1       380        40
2       390        45


calories  duration
0       420        50
1       380        40
2       390        45

## Locate Row

- Pandas Series를 반환

In [23]:
print(df.loc[0])
print(df.loc[1])
print(df.loc[2])

calories    420
duration     50
Name: 0, dtype: int64
calories    380
duration     40
Name: 1, dtype: int64
calories    390
duration     45
Name: 2, dtype: int64


- []결과는 Pandas DataFrame을 반환

In [28]:
print(df.loc[[0, 1]])
print(df.loc[[0,2]])

   calories  duration
0       420        50
1       380        40
   calories  duration
0       420        50
2       390        45


## Named Indexes

In [30]:
import pandas as pd

data = {
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}

df = pd.DataFrame(data, index = ["day1", "day2", "day3"])

print(df) 
df

      calories  duration
day1       420        50
day2       380        40
day3       390        45


calories  duration
day1       420        50
day2       380        40
day3       390        45

## Locate Named Indexes

In [31]:
print(df.loc["day2"])

calories    380
duration     40
Name: day2, dtype: int64


# Read CSV Files

In [35]:
import pandas as pd

df = pd.read_csv('./ref/data.csv')

print(df.to_string())

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.0
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

위에랑 비교하자면 toString()을 사용하면 전체 결과값을 반환하고,<br>
그냥 출력하면 head, tail 각각 5개씩만 보여준다

In [36]:
df = pd.read_csv('./ref/data.csv')

print(df)

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
..        ...    ...       ...       ...
164        60    105       140     290.8
165        60    110       145     300.0
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4

[169 rows x 4 columns]


## max_rows

In [39]:
print(pd.options.display.max_rows)

60


### 최대 행 수 늘리기 

In [43]:
import pandas as pd

pd.options.display.max_rows = 9999

df = pd.read_csv('./ref/data.csv')

# print(df)
print(pd.options.display.max_rows)

9999


# Pandas Read JSON

In [44]:
import pandas as pd

df = pd.read_json('./ref/data.json')

print(df.to_string()) 


     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
5          60    102       127     300.5
6          60    110       136     374.0
7          45    104       134     253.3
8          30    109       133     195.1
9          60     98       124     269.0
10         60    103       147     329.3
11         60    100       120     250.7
12         60    106       128     345.3
13         60    104       132     379.3
14         60     98       123     275.0
15         60     98       120     215.2
16         60    100       120     300.0
17         45     90       112       NaN
18         60    103       123     323.0
19         45     97       125     243.0
20         60    108       131     364.2
21         45    100       119     282.0
22         60    130       101     300.0
23         45   

## Dictionary as JSON

In [46]:
import pandas as pd

data = {
  "Duration":{
    "0":60,
    "1":60,
    "2":60,
    "3":45,
    "4":45,
    "5":60
  },
  "Pulse":{
    "0":110,
    "1":117,
    "2":103,
    "3":109,
    "4":117,
    "5":102
  },
  "Maxpulse":{
    "0":130,
    "1":145,
    "2":135,
    "3":175,
    "4":148,
    "5":127
  },
  "Calories":{
    "0":409,
    "1":479,
    "2":340,
    "3":282,
    "4":406,
    "5":300
  }
}

df = pd.DataFrame(data)

print(df) 
df

   Duration  Pulse  Maxpulse  Calories
0        60    110       130       409
1        60    117       145       479
2        60    103       135       340
3        45    109       175       282
4        45    117       148       406
5        60    102       127       300


Duration  Pulse  Maxpulse  Calories
0        60    110       130       409
1        60    117       145       479
2        60    103       135       340
3        45    109       175       282
4        45    117       148       406
5        60    102       127       300

# Pandas - Analyzing DataFrames

## Viewing the Data

In [47]:
import pandas as pd

df = pd.read_csv('./ref/data.csv')

print(df.head(10))

   Duration  Pulse  Maxpulse  Calories
0        60    110       130     409.1
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4
4        45    117       148     406.0
5        60    102       127     300.0
6        60    110       136     374.0
7        45    104       134     253.3
8        30    109       133     195.1
9        60     98       124     269.0


In [48]:
print(df.head())

   Duration  Pulse  Maxpulse  Calories
0        60    110       130     409.1
1        60    117       145     479.0
2        60    103       135     340.0
3        45    109       175     282.4
4        45    117       148     406.0


In [49]:
print(df.tail())

     Duration  Pulse  Maxpulse  Calories
164        60    105       140     290.8
165        60    110       145     300.0
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4


## Info About the Data

In [50]:
print(df.info()) 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 169 entries, 0 to 168
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Duration  169 non-null    int64  
 1   Pulse     169 non-null    int64  
 2   Maxpulse  169 non-null    int64  
 3   Calories  164 non-null    float64
dtypes: float64(1), int64(3)
memory usage: 5.4 KB
None


### 결과 설명

In [51]:
# <class 'pandas.core.frame.DataFrame'>
# RangeIndex: 169 entries, 0 to 168 -> 결과는 169개의 행과 4개의 열이 있다는 것을 알려줍니다.
# Data columns (total 4 columns):   -> 각 열의 이름과 데이터 유형은 다음과 같습니다.
#  #   Column    Non-Null Count  Dtype  
# ---  ------    --------------  -----  
#  0   Duration  169 non-null    int64  
#  1   Pulse     169 non-null    int64  
#  2   Maxpulse  169 non-null    int64  
#  3   Calories  164 non-null    float64
# dtypes: float64(1), int64(3)
# memory usage: 5.4 KB
# None

Null 값
- 이 info()방법은 또한 각 열에 얼마나 많은 Null이 아닌 값이 있는지 알려줍니다. <br>우리 데이터 집합에서는 "Calories" 열에 169개 중 164개의 Null이 없는 값이 있는 것으로 보입니다.
- 즉, "칼로리" 열에는 어떤 이유에서인지 값이 없는 행이 5개 있습니다.
- 빈 값 또는 Null 값은 데이터를 분석할 때 나쁠 수 있으며, 빈 값이 있는 행을 제거하는 것을 고려해야 합니다

# Pandas - Cleaning Data

- 빈 셀
- 잘못된 형식의 데이터
- 잘못된 데이터
- 중복

## Empty Cells

In [26]:
import pandas as pd

df = pd.read_csv('./ref/data2.csv')

new_df = df.dropna()
# dropna()메서드는 새 DataFrame을 반환하며 원본은 변경하지 않는다.

print(new_df.to_string())

          Duration          Date  Pulse  Maxpulse  Calories
0     0         60  '2020/12/01'    110       130     409.1
1     1         60  '2020/12/02'    117       145     479.0
2     2         60  '2020/12/03'    103       135     340.0
3     3         45  '2020/12/04'    109       175     282.4
4     4         45  '2020/12/05'    117       148     406.0
5     5         60  '2020/12/06'    102       127     300.0
6     6         60  '2020/12/07'    110       136     374.0
7     7        450  '2020/12/08'    104       134     253.3
8     8         30  '2020/12/09'    109       133     195.1
9     9         60  '2020/12/10'     98       124     269.0
10    10        60  '2020/12/11'    103       147     329.3
11    11        60  '2020/12/12'    100       120     250.7
12    12        60  '2020/12/12'    100       120     250.7
13    13        60  '2020/12/13'    106       128     345.3
14    14        60  '2020/12/14'    104       132     379.3
15    15        60  '2020/12/15'     98 

In [27]:
# NULL 값이 있는 모든 행을 제거 -> inplace = True 사용
# 원래라면 nan값이 삭제 되어서 출력되어야하는데 csv파일에 오류가 생긴 것 같음

# df = pd.read_csv('./ref/data2.csv')
df = pd.read_csv('./ref/data2.csv',  sep=",")

df.dropna(inplace = True)
# dropna(inplace = True)새 DataFrame을 반환하지 않지만 원래 DataFrame에서 NULL 값을 포함하는 모든 행을 제거.

print(df.to_string())

          Duration          Date  Pulse  Maxpulse  Calories
0     0         60  '2020/12/01'    110       130     409.1
1     1         60  '2020/12/02'    117       145     479.0
2     2         60  '2020/12/03'    103       135     340.0
3     3         45  '2020/12/04'    109       175     282.4
4     4         45  '2020/12/05'    117       148     406.0
5     5         60  '2020/12/06'    102       127     300.0
6     6         60  '2020/12/07'    110       136     374.0
7     7        450  '2020/12/08'    104       134     253.3
8     8         30  '2020/12/09'    109       133     195.1
9     9         60  '2020/12/10'     98       124     269.0
10    10        60  '2020/12/11'    103       147     329.3
11    11        60  '2020/12/12'    100       120     250.7
12    12        60  '2020/12/12'    100       120     250.7
13    13        60  '2020/12/13'    106       128     345.3
14    14        60  '2020/12/14'    104       132     379.3
15    15        60  '2020/12/15'     98 

In [28]:
print(df.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32 entries, 0 to 31
Data columns (total 1 columns):
 #   Column                                             Non-Null Count  Dtype 
---  ------                                             --------------  ----- 
 0   Duration          Date  Pulse  Maxpulse  Calories  32 non-null     object
dtypes: object(1)
memory usage: 388.0+ bytes
None


## Replace Empty Values - fillna()

In [29]:
import pandas as pd

df = pd.read_csv('./ref/data2.csv')

df.fillna(130, inplace = True) 
# 이렇게 하면 nan라고 나오는 데이터 전부가 바뀜

print(df)

    Duration          Date  Pulse  Maxpulse  Calories
0     0         60  '2020/12/01'    110       130 ...
1     1         60  '2020/12/02'    117       145 ...
2     2         60  '2020/12/03'    103       135 ...
3     3         45  '2020/12/04'    109       175 ...
4     4         45  '2020/12/05'    117       148 ...
5     5         60  '2020/12/06'    102       127 ...
6     6         60  '2020/12/07'    110       136 ...
7     7        450  '2020/12/08'    104       134 ...
8     8         30  '2020/12/09'    109       133 ...
9     9         60  '2020/12/10'     98       124 ...
10    10        60  '2020/12/11'    103       147 ...
11    11        60  '2020/12/12'    100       120 ...
12    12        60  '2020/12/12'    100       120 ...
13    13        60  '2020/12/13'    106       128 ...
14    14        60  '2020/12/14'    104       132 ...
15    15        60  '2020/12/15'     98       123 ...
16    16        60  '2020/12/16'     98       120 ...
17    17        60  '2020/12

## Replace Only For Specified Columns

In [ ]:
import pandas as pd

df = pd.read_csv('./ref/data2.csv')

df["Calories"].fillna(130, inplace = True)

Replace Using Mean, Median, or Mode

## mean() median() mode()

### 1. mean() - 평균
mean() 함수는 데이터의 산술 평균(평균값, Arithmetic Mean) 을 계산합니다.
즉, 모든 값의 합을 요소 개수로 나눈 값입니다.

In [2]:
import pandas as pd

data = pd.Series([10, 20, 30, 40, 50])
print(data.mean())

30.0


### median() - 중앙값
median() 함수는 중앙값(Median) 을 반환합니다.
모든 값을 오름차순으로 정렬했을 때 중앙에 있는 값을 의미하며, 데이터 개수가 짝수일 경우 중간 두 값의 평균을 반환합니다.

In [3]:
data = pd.Series([10, 20, 30, 40, 50])
print(data.median())  

data_even = pd.Series([10, 20, 30, 40])
print(data_even.median()) 

30.0
25.0


### mode() - 최빈값
mode() 함수는 가장 자주 등장하는 값(최빈값, Mode) 을 반환합니다.
최빈값이 여러 개일 경우, Pandas는 모든 최빈값을 반환합니다.

In [4]:
data = pd.Series([1, 2, 2, 3, 3, 3, 4, 4])
print(data.mode())

data2 = pd.Series([1, 2, 2, 3, 3])
print(data2.mode())  

0    3
dtype: int64
0    2
1    3
dtype: int64


## Pandas - Cleaning Data of Wrong Format

### to_datetime

In [ ]:
import pandas as pd

df = pd.read_csv('./ref/data2.csv')

df['Date'] = pd.to_datetime(df['Date'])

print(df.to_string())

In [ ]:
  # 22 45 NaN 100 119 282.0 
  # 23 60 '2020/12/23' 130 101 300.0 
  # 24 45 '2020/12/24' 105 132 246.0 
  # 25 60 '2020/12/25' 102 126 334.5 
  # 26 60 20201226 100 120 250.0 

# 위와 같이 nan 값이나 잘못 입력된 날짜의 형식이 

  # 22 45 NaT 100 119 282.0 
  # 23 60 '2020/12/23' 130 101 300.0 
  # 24 45 '2020/12/24' 105 132 246.0 
  # 25 60 '2020/12/25' 102 126 334.5 
  # 26 60 '2020/12/26' 100 120 250.0 

#not a time 이라는 값을 얻거나 날짜 형식대로 변환

### 행 제거

In [ ]:
# NaT 값을 제공하는데, 이는 NULL 값으로 처리할 수 있으며, 해당 dropna()메서드를 사용하여 행을 제거할 수 있습니다.
df.dropna(subset=['Date'], inplace = True)

# Pandas - Fixing Wrong Data


"잘못된 데이터"는 반드시 "빈 셀"이나 "잘못된 형식"일 필요는 없으며, 단순히 "1.99" 대신 "199"를 등록한 경우처럼 잘못된 것일 수도 있습니다.

### Replacing Values

In [ ]:
df.loc[7, 'Duration'] = 45

In [10]:
#작은 데이터 세트의 경우 잘못된 데이터를 하나씩 바꿀 수 있지만, 큰 데이터 세트의 경우 그럴 수 없습니다.
#더 큰 데이터 집합에서 잘못된 데이터를 대체하려면 몇 가지 규칙을 만들 수 있습니다. 
#예를 들어, 합법적인 값에 대한 경계를 설정하고 경계 밖에 있는 모든 값을 대체할 수 있습니다.

# ex 값이 120보다 큰 경우 120으로 설정합니다.
for x in df.index:
  if df.loc[x, "Duration"] > 120:
    df.loc[x, "Duration"] = 120

### Removing Rows

In [ ]:
for x in df.index:
  if df.loc[x, "Duration"] > 120:
    df.drop(x, inplace = True)

# Pandas - Removing Duplicates

### duplicated()
duplicated() 사용하면 중복된 값은 True, 중복되지 않은 값은 False로 나온다

In [ ]:
print(df.duplicated())

### Removing Duplicates

이 메서드(inplace = True)는 새 DataFrame을 반환하지 않지만 , 원본 DataFrame 에서 모든 중복을 제거합니다.

In [ ]:
df.drop_duplicates(inplace = True)

# Pandas - Data Correlations

## Finding Relationships

In [12]:
# corr()방법은 "숫자가 아닌" 열을 무시
import pandas as pd

df = pd.read_csv('./ref/data3.csv')
df.corr()

Duration     Pulse  Maxpulse  Calories
Duration  1.000000 -0.155408  0.009403  0.922717
Pulse    -0.155408  1.000000  0.786535  0.025121
Maxpulse  0.009403  0.786535  1.000000  0.203813
Calories  0.922717  0.025121  0.203813  1.000000

결과는 corr()두 열 사이의 관계가 얼마나 좋은지를 나타내는 많은 숫자가 포함된 표입니다.

숫자는 -1부터 1까지입니다.

1은 1 대 1 관계(완벽한 상관 관계)가 있음을 의미하고, 이 데이터 집합의 경우 첫 번째 열의 값이 올라갈 때마다 다른 열의 값도 함께 올라갑니다.

0.9도 좋은 관계이며, 한 값을 늘리면 다른 값도 늘어날 가능성이 큽니다.

-0.9는 0.9와 마찬가지로 좋은 관계이지만, 한 값을 늘리면 다른 값은 줄어들 것입니다.

0.2는 좋은 관계가 아님을 의미합니다. 즉, 한 값이 올라간다고 해서 다른 값도 올라간다는 의미는 아닙니다.

완벽한 상관관계:
"Duration"과 "Duration"이 숫자를 가지고 있는 것을 볼 수 있는데 1.000000, 이는 각 열이 항상 자체적으로 완벽한 관계를 가지고 있다는 점에서 합리적입니다.

좋은 상관관계:
"운동 시간"과 "칼로리"는 0.922721상관 관계가 있는데, 이 상관 관계는 매우 좋습니다. 운동 시간이 길수록 칼로리 소모량도 많아지고, 그 반대의 경우도 마찬가지죠. 칼로리를 많이 소모했다면 운동 시간이 길었을 가능성이 큽니다.

나쁜 상관관계:
"기간"과 "최대 펄스"는 0.009403상관관계를 가지고 있는데, 이는 매우 나쁜 상관관계입니다. 즉, 운동 기간만 보고 최대 펄스를 예측할 수 없고 그 반대의 경우도 마찬가지입니다.

# Pandas - Plotting